### PR TITLE
Add warning to prevent setting this.state to this.props referentially

### LINKED ISSUE #11593: Warn if this.state is set to this.props referentially
See this example: https://github.com/facebook/react/issues/11508#issuecomment-345464797.

I think it probably reflects a misunderstanding of how props and state work, and we should detect and warn if we see this.

### PR DESCRIPTION
**Pull request by veekas**
Adds warning to prevent setting `this.state = props`.  Closes #11593.

+ Passed all tests with `yarn test`
+ Ran `yarn prettier`
+ Ran `yarn lint`
+ Ran `yarn flow`
+ Completed CLA

Thanks for your help on this first PR, @sw-yx and @gaearon.

### CODE REVIEW

**Comment by gaearon** [comment]
Most people don't know what "referentially" means.
We also need to include the component name.

I would reword this as:

>It looks like the %s component contains a line like this in its constructor:
>
>  this.state = props;
>
>This is not recommended because any further updates to props won't be reflected in the state. In most cases, you don't need to keep state and props in sync. Instead, use props directly. If you need to calculate something from the props, do it during the rendering. If you need to share the state between several components, move it to their closest common ancestor, and pass it down as props to them.

**Comment by veekas** [comment]
Updated to include the new warning message and component name. Ran `test`/`prettier`/`linc`/`flow` again.

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 1]
```diff
@@ -465,6 +465,19 @@ export default function(
     instance.refs = emptyObject;
     instance.context = getMaskedContext(workInProgress, unmaskedContext);
 
+    const noSetStatesFromProps = !(instance.state === instance.props);
+    warning(
+      noSetStatesFromProps,
```
Comment (aweary) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: I know this is consistent with the rest of the warnings, but I wonder if we should just inline `noSetStatesFromProps` and avoid the whole negation thing?

```js
warning(
  instance.state !== instance.props,
  `...`
)
```

Comment (veekas) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: That looks better to me, as well. Should I make the edit?

Comment (sw-yx) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: the guidance is not clear here tbh. i've seen @gaearon also prefer `warning(false,....` and doing all the logic in an if block. ¯\_(ツ)_/¯

Comment (aweary) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: Typically we do `warning(false,  "")` when we also need to execute some other code when issuing the warning, such as setting a flag so that the warning is deduped.

**Comment by veekas** [comment]
Hi @gaearon and @aweary, I wanted to check on this issue. Is there anything else you would like me to do to help close it?

File: packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js [hunk 2]
```diff
@@ -197,6 +197,32 @@ describe('ReactComponentLifeCycle', () => {
...
+      render() {
+        return <div />;
+      }
+    }
+
+    ReactTestUtils.renderIntoDocument(<StatefulComponent />);
+    expectDev(console.error.calls.count()).toBe(1);
+    expectDev(console.error.calls.argsFor(0)[0]).toContain(
+      'It looks like the StatefulComponent component contains a line like this in its constructor: ' +
+        'this.state = props; ' +
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js]: Have you checked how this looks in a browser? Since you didn't add the linebreaks I assume it will look a bit confusing.

Comment (veekas) [reply packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js]: Good point. Pushed a new commit with line breaks before and after the code excerpt to increase readability.

**Comment by veekas** [comment]
Looks like the build is failing due to the same CI/Docker issue (nodejs/docker-node#651) referenced by @bvaughn in #12392. Lint, flow, and tests are passing locally.

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 3]
```diff
@@ -620,6 +620,20 @@ export default function(
...
+      'It looks like the %s component contains a line like this ' +
+        'in its constructor:\n\n' +
+        'this.state = props;\n\n' +
+        "This is not recommended because any further updates to props won't be " +
+        "reflected in the state. In most cases, you don't need to keep state and " +
+        'props in sync. Instead, use the props directly. If you need to calculate ' +
+        'something from the props, do it during the rendering. If you need to ' +
+        'share the state between several components, move it to their closest ' +
+        'common ancestor and pass it down as props to them.',
+      getComponentName(workInProgress) || 'Unknown',
```
Comment (raunofreiberg) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: Just a small nit: it would probably be a better idea to pass `a component` instead of `Unknown` if no component name is found.

**Your example:**

```
It looks like the Unknown component contains a line ...
```

**Suggestion:**
```
A component contains a line ...
```

or if the component name exists

```
Foo contains a line like this...
```

This also would follow the consistency of handling missing component names in error messages. See: [ReactFiberTreeReflection.js](https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberTreeReflection.js#L77), [ReactDOM.js](https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOM.js#L1139)

Comment (veekas) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: I'm inclined to agree, but as I searched through the code base I saw the `getComponentName(*) || 'Unknown'` pattern used 22 times while `Component` or `A component` were only used 7 and 8 times respectively. In fact, the most recent commits by @gaearon, @acdlite, and @bvaughn all use `Unknown` (#12028, #12359, #11455). I'll let one of them chime in about whether they think this should be changed.

Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: It's true that "Unknown" is probably a more common fallback name, but it's not strictly necessary to use that pattern. That being said, I think this warning is a bit verbose and maybe tries to cover too broad a topic (e.g. sharing state between several components).

At first glance, I would prefer a shorter message, something like:
```js
warning(
  instance.state !== instance.props,
  "%: It is not recommended to assign props directly to state. " +
    "Updates to props won't be reflected in state. " +
    "It's usually better to just use props directly.",
  getComponentName(workInProgress) || "Unknown"
);
```

**Comment by veekas** [comment]
It's been a couple weeks without comment, so I incorporated @bvaughn's and @raunofreiberg's input into the latest iteration. Thanks, you two, I think your edits make it more semantic and terse. Pinging @gaearon

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 4]
```diff
@@ -700,6 +700,14 @@ export default function(
     instance.refs = emptyObject;
     instance.context = getMaskedContext(workInProgress, unmaskedContext);
 
+    warning(
+      instance.state !== instance.props,
+      '%s: It is not recommended to assign props directly to state' +
+        "because updates to props won't be reflected in state." +
+        'In most cases, it is better to use props directly.',
+      getComponentName(workInProgress) || 'A component',
+    );
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: This needs to be within an `if (__DEV__)` check. Suggest using the one right below (line 711).

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 5]
```diff
@@ -700,6 +700,14 @@ export default function(
     instance.refs = emptyObject;
     instance.context = getMaskedContext(workInProgress, unmaskedContext);
 
+    warning(
+      instance.state !== instance.props,
+      '%s: It is not recommended to assign props directly to state' +
+        "because updates to props won't be reflected in state." +
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: Need a trailing space (" ") here.

Comment (veekas) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: Fixed!

File: packages/react-reconciler/src/ReactFiberClassComponent.js (hunk 4 above)
Comment (veekas) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: Fixed!

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 6]
```diff
@@ -700,15 +700,15 @@ export default function(
...
-      getComponentName(workInProgress) || 'A component',
-    );
-
     if (__DEV__) {
+      warning(
+        instance.state !== instance.props,
+        '%s: It is not recommended to assign props directly to state ' +
+          "because updates to props won't be reflected in state. " +
+          'In most cases, it is better to use props directly.',
+        getComponentName(workInProgress) || 'A component',
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: nit: I think we generally use "Component" as the default name in warnings when it's followed by ":" (e.g. "Component: blah". We use "A component" when it's part of a sentence (e.g. "A component blah")

**Comment by bvaughn** [comment]
Nice catch 😄

**Comment by veekas** [comment]
Latest commit should address @gaearon's concerns!

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 7]
```diff
@@ -700,6 +702,18 @@ export default function(
     instance.context = getMaskedContext(workInProgress, unmaskedContext);
 
     if (__DEV__) {
+      const componentName = getComponentName(workInProgress) || 'Component';
+      if (!didWarnAboutDirectlyAssigningPropsToState.has(componentName)) {
+        didWarnAboutDirectlyAssigningPropsToState.add(componentName);
+        warning(
+          instance.state !== instance.props,
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: I think we should move this check up a level:
```js
if (__DEV__) {
  if (instance.state === instance.props) {
    const componentName = getComponentName(workInProgress) || 'Component';
    if (!didWarnAboutDirectlyAssigningPropsToState.has(componentName)) {
      didWarnAboutDirectlyAssigningPropsToState.add(componentName);
      warning(false, ...);
```
This way we don't add things to the Set unnecessarily, and we also don't potentially suppress a later warning (e.g. if we encounter 2 components with the same name- and the second one assigns props to state).

Comment (veekas) [reply packages/react-reconciler/src/ReactFiberClassComponent.js]: Ah, yes, of course. Thanks for catching that.

**Comment by react-linter** [comment]
<details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 9c77ffb444598c32c8f92c8d79e406959a10445b...1a7463d7bd866066c21e947fcaf684d4feb298cb</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | +0.1% | 611.19 KB | 611.8 KB | 141.28 KB | 141.38 KB | UMD_DEV
react-dom.development.js | +0.1% | +0.1% | 595.56 KB | 596.17 KB | 137.14 KB | 137.24 KB | NODE_DEV
ReactDOM-dev.js | +0.1% | +0.1% | 619.99 KB | 620.68 KB | 139.85 KB | 139.96 KB | FB_WWW_DEV

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.1% | +0.1% | 414.1 KB | 414.72 KB | 90.15 KB | 90.25 KB | UMD_DEV
react-art.development.js | +0.2% | +0.1% | 339.94 KB | 340.55 KB | 71.51 KB | 71.61 KB | NODE_DEV
ReactART-dev.js | +0.2% | +0.1% | 348.2 KB | 348.89 KB | 71.06 KB | 71.16 KB | FB_WWW_DEV

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +0.2% | +0.1% | 348.03 KB | 348.65 KB | 73.21 KB | 73.31 KB | UMD_DEV
react-test-renderer.development.js | +0.2% | +0.1% | 338.86 KB | 339.47 KB | 70.53 KB | 70.63 KB | NODE_DEV
ReactTestRenderer-dev.js | +0.2% | +0.1% | 347.4 KB | 348.08 KB | 70.15 KB | 70.25 KB | FB_WWW_DEV

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +0.2% | +0.2% | 319.2 KB | 319.81 KB | 65.98 KB | 66.08 KB | NODE_DEV
react-reconciler-persistent.development.js | +0.2% | +0.2% | 318.53 KB | 319.14 KB | 65.75 KB | 65.85 KB | NODE_DEV

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +0.1% | +0.1% | 459.01 KB | 459.7 KB | 98.2 KB | 98.3 KB | RN_FB_DEV
ReactNativeRenderer-dev.js | +0.1% | +0.1% | 458.76 KB | 459.45 KB | 98.14 KB | 98.24 KB | RN_OSS_DEV
ReactFabric-dev.js | +0.2% | +0.1% | 441.43 KB | 442.11 KB | 93.8 KB | 93.9 KB | RN_FB_DEV
ReactFabric-dev.js | +0.2% | +0.1% | 441.46 KB | 442.15 KB | 93.82 KB | 93.92 KB | RN_OSS_DEV

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by veekas** [comment]
Hey @gaearon, just want to ping you again about this PR

**Comment by veekas** [comment]
Hi @gaearon, I've rebased/squashed my commits to fix the merge conflict that happened over the past few months. I'll give this opportunity to one of the volunteers in #11593 if any further changes are requested. Thanks!

**Comment by gaearon** [comment]
Thanks!

**Comment by stnwk** [comment]
Is there any reason this is being outputted as an `error` on the console instead of a `warning`? @veekas @gaearon

**Comment by gaearon** [comment]
@stnwk We output all warnings through `console.error` because React warnings generally point out legitimate bugs in your code. If you're 100% confident you know why you're copying props to state like this, you can change it to `this.state = {...this.props}` to be explicit. But it's usually a mistake.

**Comment by stnwk** [comment]
Okay, thanks for the explanation :)
